File: paas-metasvr/src/main/java/com/zzstack/paas/underlying/metasvr/bean/PaasRedisCluster.java
package com.zzstack.paas.underlying.metasvr.bean;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import com.zzstack.paas.underlying.utils.consts.CONSTS;

public class PaasRedisCluster {
	
    private Map<String, PaasRedisNode> masterNodes;
    private Map<String, PaasRedisNode> slaveNodes;
    
	public PaasRedisCluster() {
		super();
		masterNodes = new HashMap<String, PaasRedisNode>();
		slaveNodes = new HashMap<String, PaasRedisNode>();
	}

	public void parse(String info) {
		String[] vs = info.split("\n");

		for (String str : vs) {
			int idxAt = str.indexOf("@");
			if (idxAt == CONSTS.STR_NOT_FOUND) {
				continue;
			}
			
			if (str.charAt(str.length() - 1) == '\r') {
			    str = str.substring(0,  str.length() - 2);
			}

			// f3607081ae714026d94b4a3a74edaf5981aaf91b 192.168.31.215:8104@18104 slave
			// 171c0a4f7ed5486a399f13908f97512a86105b4b 0 1597899045000 2 connected
			// 171c0a4f7ed5486a399f13908f97512a86105b4b 192.168.31.215:8101@18101 master - 0
			// 1597899047000 2 connected 5461-10922
			// ba9e95e78c6fee2829b63f9c29e1e6a365d4ba64 192.168.31.215:8100@18100
			// myself,master - 0 1597899047000 1 connected 0-5460
			PaasRedisNode node = new PaasRedisNode();
			int idxAddr = str.indexOf(" ", 0);
			node.setNodeId(str.substring(0, idxAddr));

			int idxSem = str.indexOf(":", 0);
			String ip = str.substring(idxAddr + 1, idxSem);
			String port = str.substring(idxSem + 1, idxAt);
			node.setIp(ip);
			node.setPort(port);

			int idxMaster = str.indexOf("master", idxAt);
			int idxSlave = str.indexOf("slave", idxAt);

			if (idxMaster != CONSTS.STR_NOT_FOUND) {
				node.setRedisRole(CONSTS.REDIS_ROLE_MASTER);

				int idxSlot = str.indexOf("connected ", idxMaster);
				if (idxSlot != CONSTS.STR_NOT_FOUND) {
					String slotRange = str.substring(idxSlot + "connected ".length());
					node.setSlotRange(slotRange);
				}

				masterNodes.put(node.getNodeId(), node);
			} else if (idxSlave != CONSTS.STR_NOT_FOUND) {
				node.setRedisRole(CONSTS.REDIS_ROLE_SLAVE);

				int idxIdBeg = idxSlave + "slave ".length();
				int idxIdEnd = str.indexOf(" ", idxIdBeg);

				String masterId = str.substring(idxIdBeg, idxIdEnd);
				node.setMasterId(masterId);

				slaveNodes.put(node.getNodeId(), node);
			}

			// String debug = String.format("role:%d, port:%s, slot:%s",
			// node.getRedisRole(), node.getPort(), node.getSlotRange());
			// System.out.println(debug);
		}

		Set<Entry<String, PaasRedisNode>> entrySet = slaveNodes.entrySet();
		for (Entry<String, PaasRedisNode> entry : entrySet) {
		    PaasRedisNode slave = entry.getValue();
			String masterId = slave.getMasterId();

			PaasRedisNode master = masterNodes.get(masterId);
			if (master != null) {
				master.addSlaveId(slave.getNodeId());
			}
		}
	}
	
	public int getMasterCount() {
		int cnt = 0;
		Set<Entry<String, PaasRedisNode>> entrySet = masterNodes.entrySet();
		for (Entry<String, PaasRedisNode> entry : entrySet) {
		    PaasRedisNode master = entry.getValue();
			if (!master.getNodeId().isEmpty() && !master.getSlotRange().isEmpty()) {
				++cnt;
			}
		}
		return cnt;
	}
	
	public String getAloneMaster() {
		String ret = "";

		Set<Entry<String, PaasRedisNode>> entrySet = masterNodes.entrySet();
		for (Entry<String, PaasRedisNode> entry : entrySet) {
		    PaasRedisNode master = entry.getValue();
			if (master.isSlaveEmpty()) {
				ret = master.getNodeId();
				break;
			}
		}
		return ret;
	}
	
	public String getOneMasterAddr() {
		StringBuilder addr = new StringBuilder("");

		Set<Entry<String, PaasRedisNode>> entrySet = masterNodes.entrySet();
		for (Entry<String, PaasRedisNode> entry : entrySet) {
		    PaasRedisNode master = entry.getValue();
			if (!master.getIp().isEmpty() && !master.getPort().isEmpty()) {
				addr.append(master.getIp()).append(":").append(master.getPort());
				break;
			}
		}
		return addr.toString();
	}
	
	public String getMasterIdList() {
		StringBuilder masterIds = new StringBuilder("");
		boolean filled = false;
		Set<Entry<String, PaasRedisNode>> entrySet = masterNodes.entrySet();
		for (Entry<String, PaasRedisNode> entry : entrySet) {
		    PaasRedisNode master = entry.getValue();
			if (!master.getNodeId().isEmpty() && !master.getSlotRange().isEmpty()) {
				if (filled)
					masterIds.append(",");
				masterIds.append(master.getNodeId());

				if (!filled)
					filled = true;
			}
		}
		return masterIds.toString();
	}
	
	public PaasRedisNode getSelfInfo(String ip, String port) {
		Set<Entry<String, PaasRedisNode>> entrySet = masterNodes.entrySet();
		for (Entry<String, PaasRedisNode> entry : entrySet) {
			PaasRedisNode node = entry.getValue();
			if (node.getIp().equals(ip) && node.getPort().equals(port)) {
				return node;
			}
		}

		entrySet = slaveNodes.entrySet();
		for (Entry<String, PaasRedisNode> entry : entrySet) {
			PaasRedisNode node = entry.getValue();
			if (node.getIp().equals(ip) && node.getPort().equals(port)) {
				return node;
			}
		}

		return null;
	}
	
	public void getSlaves(String masterId, ArrayList<PaasRedisNode> slaves) {	    
	    PaasRedisNode master = masterNodes.get(masterId);
	    if (master != null) {
	        ArrayList<String> slaveIds = master.getSlaveIds();
	        for (String slaveId : slaveIds) {
	            PaasRedisNode slave = slaveNodes.get(slaveId);
	            slaves.add(slave);
	        }
	    }
	}
	
    public void getSlotedMasters(ArrayList<PaasRedisNode> masters) {
        Set<Entry<String, PaasRedisNode>> entrySet = masterNodes.entrySet();
        for (Entry<String, PaasRedisNode> entry : entrySet) {
            masters.add(entry.getValue());
        }
    }
    
    public int getNodeSlotCount(String nodeId) {
        PaasRedisNode node = masterNodes.get(nodeId);
        if (node == null) return 0;
        
        String slotRange = node.getSlotRange();
        
        // 0-1364 5461-6825 10923-12287
        String[] vs = slotRange.split(" ");
        int slotCnt = 0;
        for (int i = 0; i < vs.length; ++i) {
            String str = vs[i];
            if (str.isEmpty()) continue;
            
            int splitIdx = str.indexOf("-");
            if (splitIdx == -1) slotCnt += 1;

            String sStart = str.substring(0, splitIdx);
            String sEnd   = str.substring(splitIdx + 1);

            int start = Integer.valueOf(sStart);
            int end   = Integer.valueOf(sEnd);
            int range = end - start + 1;
            slotCnt += range;
        }
        
        return slotCnt;
    }

}

File: paas-metasvr-go/pkg/proto/paas_redis_cluster.go
package proto

import (
	"strings"

	"github.com/maoge/paas-metasvr-go/pkg/consts"
)

type PaasRedisCluster struct {
	MasterNodes map[string]*PaasRedisNode
	SlaveNodes  map[string]*PaasRedisNode
}

func NewPaasRedisCluster() *PaasRedisCluster {
	paasRedisCluster := new(PaasRedisCluster)
	paasRedisCluster.MasterNodes = make(map[string]*PaasRedisNode)
	paasRedisCluster.SlaveNodes = make(map[string]*PaasRedisNode)
	return paasRedisCluster
}

func (h *PaasRedisCluster) Parse(info string) {
	vs := strings.Split("info", "\n")

	for _, str := range vs {
		idxAt := strings.Index(str, "@")
		if idxAt == -1 {
			continue
		}

		size := len(str)
		if str[size-1:] == "\r" {
			str = str[:size-2]
		}

		// d5d27d0379b28a1c0965d5ee3fdead018e204c9a 172.20.0.171:13002@23002 master - 0 1649230545419 2 connected 5461-10922
		// b91f8eec7570e89c8f24ce97f2cb5d18c588415c 172.20.0.171:13001@23001 myself,master - 0 1649230545000 1 connected 0-5460
		// e1d627314e5aee042801f4a3fbd24503e8b41415 172.20.0.171:13003@23003 master - 0 1649230544000 3 connected 10923-16383
		// dd73ea25bab1aa271b93786a660cd927c924fd34 172.20.0.171:13004@23004 slave d5d27d0379b28a1c0965d5ee3fdead018e204c9a 0 1649230543000 4 connected
		// 992e816a7bc2d65443b60270f3d03173a62d492e 172.20.0.171:13005@23005 slave e1d627314e5aee042801f4a3fbd24503e8b41415 0 1649230546419 5 connected
		// 5226b1db08fe711d9dc24b5d1c09c09a9a669564 172.20.0.171:13006@23006 slave b91f8eec7570e89c8f24ce97f2cb5d18c588415c 0 1649230544416 6 connected

		idxAddr := strings.Index(str, " ")
		idxSem := strings.Index(str, ":")

		nodeId := str[:idxAddr]
		ip := str[idxAddr+1 : idxSem]
		port := str[idxSem+1 : idxAt]

		node := NewPaasRedisNode(nodeId, ip, port)

		if strings.Index(str, "master") != -1 {
			node.RedisRole = consts.REDIS_ROLE_MASTER

			idxSlot := strings.Index(str, "connected ")
			if idxSlot != -1 {
				start := idxSlot + len("connected ")
				node.SlotRange = str[start:]
			}

			h.MasterNodes[node.NodeId] = node
		} else {
			node.RedisRole = consts.REDIS_ROLE_SLAVE

			idxIdBeg := strings.Index(str, "slave ") + len("slave ")
			subStr := str[idxIdBeg:]
			idxIdEnd := strings.LastIndex(subStr, " ")

			masterId := subStr[:idxIdEnd]
			node.MasterId = masterId

			h.SlaveNodes[node.NodeId] = node
		}

	}

	for _, slave := range h.SlaveNodes {
		master, found := h.MasterNodes[slave.MasterId]
		if found {
			master.AddSlaveId(slave.NodeId)
		}
	}
}

func (h *PaasRedisCluster) GetSelfInfo(ip, port string) *PaasRedisNode {
	for _, node := range h.MasterNodes {
		if node.Ip == ip && node.Port == port {
			return node
		}
	}

	for _, node := range h.SlaveNodes {
		if node.Ip == ip && node.Port == port {
			return node
		}
	}

	return nil
}

func (h *PaasRedisCluster) GetSlaves(masterId string) []*PaasRedisNode {
	result := make([]*PaasRedisNode, 0)
	master, found := h.MasterNodes[masterId]
	if !found {
		return result
	}

	slaveIds := master.SlaveIds
	for _, slaveId := range slaveIds {
		slave := h.SlaveNodes[slaveId]
		result = append(result, slave)
	}

	return result
}

func (h *PaasRedisCluster) GetMasters() []*PaasRedisNode {
	result := make([]*PaasRedisNode, 0)
	for _, node := range h.MasterNodes {
		result = append(result, node)
	}

	return result
}

func (h *PaasRedisCluster) GetAloneMaster() string {
	for _, node := range h.MasterNodes {
		if node.IsSlaveEmpty() {
			return node.NodeId
		}
	}
	return ""
}

func (h *PaasRedisCluster) GetAnyMasterAddr() (string, bool) {
	for _, master := range h.MasterNodes {
		return master.Ip + ":" + master.Port, true
	}

	return "", false
}
